File: src/persistence.js
import Database from 'better-sqlite3'

export class Persistence {
  constructor(dbPath = 'rootspace.db') {
    this.db = new Database(dbPath)
    this.init()
  }

  init() {
    // Create messages table
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS messages (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        peer_id TEXT NOT NULL,
        topic TEXT NOT NULL,
        data TEXT NOT NULL,
        signature TEXT,
        public_key TEXT,
        timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
      );

      CREATE TABLE IF NOT EXISTS peers (
        peer_id TEXT PRIMARY KEY,
        reputation INTEGER DEFAULT 100,
        last_seen DATETIME DEFAULT CURRENT_TIMESTAMP
      );
    `)
    console.log('Persistence: SQLite database initialized.')
  }

  saveMessage({ peerId, topic, data, signature, publicKey }) {
    const stmt = this.db.prepare(`
      INSERT INTO messages (peer_id, topic, data, signature, public_key)
      VALUES (?, ?, ?, ?, ?)
    `)
    return stmt.run(peerId, topic, JSON.stringify(data), signature, publicKey)
  }

  getHistory(limit = 100) {
    const stmt = this.db.prepare('SELECT * FROM messages ORDER BY timestamp DESC LIMIT ?')
    return stmt.all(limit)
  }

  updateReputation(peerId, delta) {
    // Ensure peer exists
    this.db.prepare('INSERT OR IGNORE INTO peers (peer_id) VALUES (?)').run(peerId)

    const stmt = this.db.prepare(`
      UPDATE peers 
      SET reputation = reputation + ?, last_seen = CURRENT_TIMESTAMP 
      WHERE peer_id = ?
    `)
    return stmt.run(delta, peerId)
  }

  getPeers() {
    return this.db.prepare('SELECT * FROM peers ORDER BY reputation DESC').all()
  }
}

File: rust-v2/src/persistence.rs
use rusqlite::{Connection, Result, params};
use std::path::Path;

pub struct Persistence {
    conn: Connection,
}

impl Persistence {
    pub fn new<P: AsRef<Path>>(path: P) -> Result<Self> {
        let conn = Connection::open(path)?;
        Self::init(&conn)?;
        Ok(Self { conn })
    }

    fn init(conn: &Connection) -> Result<()> {
        conn.execute(
            "CREATE TABLE IF NOT EXISTS messages (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                peer_id TEXT NOT NULL,
                topic TEXT NOT NULL,
                data TEXT NOT NULL,
                timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
            )",
            [],
        )?;

        conn.execute(
            "CREATE TABLE IF NOT EXISTS audit_logs (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                event_type TEXT NOT NULL,
                actor TEXT NOT NULL,
                action TEXT NOT NULL,
                result TEXT NOT NULL,
                timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
            )",
            [],
        )?;
        Ok(())
    }

    pub fn save_message(&self, peer_id: &str, topic: &str, data: &str) -> Result<()> {
        self.conn.execute(
            "INSERT INTO messages (peer_id, topic, data) VALUES (?, ?, ?)",
            params![peer_id, topic, data],
        )?;
        Ok(())
    }

    pub fn log_audit_event(&self, event_type: &str, actor: &str, action: &str, result: &str) -> Result<()> {
        self.conn.execute(
            "INSERT INTO audit_logs (event_type, actor, action, result) VALUES (?, ?, ?, ?)",
            params![event_type, actor, action, result],
        )?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_audit_logging() -> Result<()> {
        let db = Persistence::new(":memory:")?;
        db.log_audit_event("TEST", "ACTOR", "ACTION", "SUCCESS")?;

        let mut stmt = db.conn.prepare("SELECT event_type, actor, action, result FROM audit_logs")?;
        let mut rows = stmt.query([])?;

        if let Some(row) = rows.next()? {
            let event_type: String = row.get(0)?;
            let actor: String = row.get(1)?;
            let action: String = row.get(2)?;
            let result: String = row.get(3)?;

            assert_eq!(event_type, "TEST");
            assert_eq!(actor, "ACTOR");
            assert_eq!(action, "ACTION");
            assert_eq!(result, "SUCCESS");
        } else {
            panic!("No audit log entry found");
        }

        Ok(())
    }
}
